### ISSUE TITLE
`language-c99-simple-0.2.0.0` regression involving initializer lists

### ISSUE THREAD

**Issue by RyanGlScott** [fdedden/language-c99-simple#8]
_(Originally reported at https://github.com/Copilot-Language/copilot/issues/335#issuecomment-1175087673.)_

After upgrading to `language-c99-simple-0.2.0.0`, I noticed that array initializers were being pretty-printed in a different order than what older versions of the library would print. Here is a minimal example that demonstrates the issue, with some light CPP used to paper over API differences between the two major versions of the library:

```hs
#!/usr/bin/env cabal
{- cabal:
build-depends: base, language-c99, language-c99-simple
-}
{-# LANGUAGE CPP #-}
module Main where

import GHC.Exts (IsList(..))
import Language.C99.Pretty
import Language.C99.Simple

arr :: Init
arr =
#if MIN_VERSION_language_c99_simple(0,2,0)
  InitList $
#else
  InitArray $
#endif
  fromList [ boolInit True
           , boolInit False
           , boolInit False
           , boolInit False
           , boolInit False
           ]

#if MIN_VERSION_language_c99_simple(0,2,0)
boolInit :: Bool -> InitItem
boolInit = InitItem Nothing . InitExpr . LitBool
#else
boolInit :: Bool -> Init
boolInit = InitExpr . LitBool
#endif

main :: IO ()
main = print $ pretty $ transinit arr
```

Running this program with `language-c99-simple-0.1.*` gives this order, as expected:

```
$ cabal run Bug.hs --constraint="language-c99-simple<0.2"
<snip>
{(true), (false), (false), (false), (false)}
```

On the other hand, running this program with `language-c99-simple-0.2.*` gives an unexpected order:

```
$ cabal run Bug.hs --constraint="language-c99-simple>=0.2"
<snip>
{(false), (false), (false), (false), (true)}
```

Notice that the `true` element is now at the end of the array instead of the beginning. Bisecting reveals that commit 86fe2b374036204c954dd11805f981ea8e233813 is responsible for this regression.

**Comment by RyanGlScott** [comment]
I believe this is happening because prior to 86fe2b374036204c954dd11805f981ea8e233813, the translation from `language-c99-simple` to `language-c99` was using `fromList` to construct the `InitList`:

```hs
transinit (InitArray es) = C.InitArray (fromList $ map transinit es)
```

Note that the [`IsList InitList` instance](https://github.com/fdedden/language-c99-util/blob/e5a123b27867e5a235d6ab8ea5217b18b6faf621/src/Language/C99/Util/IsList.hs#L30-L35) will use a left fold to construct the `InitList`. This makes sense, since `InitList` is a snoc-list rather than a cons-list.

After 86fe2b374036204c954dd11805f981ea8e233813, however, `fromList` is no longer used to construct the `InitList`. Instead, it is constructed using an auxiliary `transinitlist` function:

```hs
transinit (InitMultiple es) = C.InitArray (transinitlist es)

transinitlist :: NonEmpty InitItem -> C.InitList
transinitlist (InitItem mident x NE.:| [])     = C.InitBase
  (transdesigr <$> mident) (transinit x)
transinitlist (InitItem mident x NE.:| (y:zs)) = C.InitCons
  (transinitlist (y NE.:| zs)) (transdesigr <$> mident) (transinit x)
```

This uses a _right_ fold rather than a left fold, which I believe is responsible for the list of elements being reversed. To fix the bug, it would suffice to switch `transinitlist` to use a left fold instead of a right fold. Something like this would work:

```diff
diff --git a/src/Language/C99/Simple/Translate.hs b/src/Language/C99/Simple/Translate.hs
index b66d6ae..a2a3807 100644
--- a/src/Language/C99/Simple/Translate.hs
+++ b/src/Language/C99/Simple/Translate.hs
@@ -278,10 +278,10 @@ transinit (InitExpr e)  = C.InitExpr (wrap $ transexpr e)
 transinit (InitList es) = C.InitList (transinitlist es)
 
 transinitlist :: NonEmpty InitItem -> C.InitList
-transinitlist (InitItem mident x NE.:| [])     = C.InitBase
-  (transdesigr <$> mident) (transinit x)
-transinitlist (InitItem mident x NE.:| (y:zs)) = C.InitCons
-  (transinitlist (y NE.:| zs)) (transdesigr <$> mident) (transinit x)
+transinitlist (x NE.:| xs) = foldl step (base x) xs
+  where
+    base (InitItem mident y)    = C.InitBase    (transdesigr <$> mident) (transinit y)
+    step ys (InitItem mident y) = C.InitCons ys (transdesigr <$> mident) (transinit y)
 
 transdesigr :: Ident -> C.Design
 transdesigr = C.Design . C.DesigrBase . C.DesigrIdent . ident
@@ -292,10 +292,7 @@ initexpr ty inits = C.PostfixInits ty' inits' where
   inits' = transinititems inits
 
 transinititems :: NonEmpty InitItem -> C.InitList
-transinititems (InitItem mident init NE.:| []) =
-  C.InitBase ((C.Design . C.DesigrBase . C.DesigrIdent . ident) <$> mident) (transinit init)
-transinititems (InitItem mident init NE.:| (x:xs)) =
-  C.InitCons (transinititems (x NE.:| xs)) ((C.Design . C.DesigrBase . C.DesigrIdent . ident) <$> mident) (transinit init)
+transinititems = transinitlist
 
 indexexpr arr idx = C.PostfixIndex arr' idx' where
   arr' = wrap $ transexpr arr
```

**Comment by RyanGlScott** [comment]
Now that I look closer, there is an analogous bug that was introduced in 12895af2ee5ff86ad1d45368307e05664eae5161, this time involving struct and enum initializers. Again, that commit removes `fromList`, a left fold, in favor of right folds. To fix those bugs, we'd need corresponding left folds for structs and enums as well:

```diff
diff --git a/src/Language/C99/Simple/Translate.hs b/src/Language/C99/Simple/Translate.hs
index a2a3807..db82298 100644
--- a/src/Language/C99/Simple/Translate.hs
+++ b/src/Language/C99/Simple/Translate.hs
@@ -168,10 +168,10 @@ spec2spec ts = case ts of
     declns' = transvariantdeclns declns
 
 transfielddeclns :: NonEmpty FieldDecln -> C.StructDeclnList
-transfielddeclns (decln NE.:| []) = C.StructDeclnBase (transfielddecln decln)
-transfielddeclns (decln NE.:| (d : ds)) = C.StructDeclnCons
-  (transfielddeclns (d NE.:| ds))
-  (transfielddecln decln)
+transfielddeclns (decln NE.:| declns) = foldl step (base decln) declns
+  where
+    base d    = C.StructDeclnBase (transfielddecln d)
+    step ds d = C.StructDeclnCons ds (transfielddecln d)
 
 transfielddecln :: FieldDecln -> C.StructDecln
 transfielddecln (FieldDecln ty name) = C.StructDecln quals declrlist where
@@ -180,10 +180,10 @@ transfielddecln (FieldDecln ty name) = C.StructDecln quals declrlist where
   quals = getspecquals ty
 
 transvariantdeclns :: NonEmpty Ident -> C.EnumrList
-transvariantdeclns (decln NE.:| []) = C.EnumrBase (transvariantdecln decln)
-transvariantdeclns (decln NE.:| (d : ds)) = C.EnumrCons
-  (transvariantdeclns (d NE.:| ds))
-  (transvariantdecln decln)
+transvariantdeclns (decln NE.:| declns) = foldl step (base decln) declns
+  where
+    base d    = C.EnumrBase (transvariantdecln d)
+    step ds d = C.EnumrCons ds (transvariantdecln d)
 
 transvariantdecln :: Ident -> C.Enumr
 transvariantdecln name = C.Enumr (C.Enum (ident name))
```

**Comment by fdedden** [comment]
Thanks @RyanGlScott for pointing this out. This is definitely a bug, clearly some failure on my side regarding testing.

**Comment by RyanGlScott** [comment]
Thanks, @fdedden!

Would you be willing to make a bugfix release of `language-c99-simple` to Hackage with these changes? Or is that waiting on a resolution to #7 first?

**Comment by fdedden** [comment]
You're welcome! I will make a new release right now, as I don't know when a new version of mtl will be available. Once that is public, I can always do a new point release with the updated dependency.

**Comment by fdedden** [comment]
I have published `0.2.1` on hackage, and deprecated `0.2.0` (https://hackage.haskell.org/package/language-c99-simple). I have updated Copilot to work with `>= 0.2.0` as well, which should be included in the next release.

**Comment by RyanGlScott** [comment]
Thanks again!
